[16.0][ADD] stock_picking_backorder_reason

rousseldenis opened the pull request:
In some flows, one may want to let the user have the choice on something more understandable in his context. So, instead choosing a backorder strategy, let him choose a 'reason'.

Let define some backorder reasons and associate the strategy on those reasons.

We define also a new strategy which is 'Use partner option' to delegate it on partner level.

rousseldenis committed:
[16.0][ADD] stock_picking_backorder_reason

lmignon commented on stock_picking_backorder_reason/models/stock_backorder_reason.py:
```diff
@@ -0,0 +1,22 @@
+            ("cancel", "Cancel backorder"),
+            ("use_partner_option", "Use partner option"),
+        ],
+        default="picking_type",
```
@rousseldenis `picking_type` is not into the selection list

rousseldenis replied:
Done

lmignon commented on stock_picking_backorder_reason/demo/stock_backorder_reason.xml:
```diff
@@ -0,0 +1,12 @@
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo noupdate="1">
+
+    <!-- TODO
+    <record model="stock.backorder.reason" id="stock_backorder_reason_demo_1">
+       <field name="name">...</field>
+    </record>
+    -->
```
@rousseldenis Do we need-it?

rousseldenis replied:
I've added some

lmignon commented on stock_picking_backorder_reason/models/stock_picking_type.py:
```diff
@@ -0,0 +1,57 @@
+
+    @api.depends("backorder_reason")
+    def _compute_backorder_reason_sale(self):
+        activated_types = self.filtered("backorder_reason_sale")
+        to_deactivate = activated_types.filtered(lambda t: not t.backorder_reason)
```
a double loop.... should be done in one loop.

rousseldenis replied:
Done

lmignon commented on stock_picking_backorder_reason/models/stock_picking_type.py:
```diff
@@ -0,0 +1,57 @@
+
+    @api.depends("backorder_reason")
+    def _compute_backorder_reason_purchase(self):
+        activated_types = self.filtered("backorder_reason_purchase")
+        to_deactivate = activated_types.filtered(lambda t: not t.backorder_reason)
```
a double loop.... should be done in one loop.

rousseldenis committed:
[IMP] stock_picking_backorder_reason: Remove several loops

rousseldenis committed:
[IMP] stock_picking_backorder_reason: Allow to use picking type behavior + several fixes

rousseldenis committed:
[IMP] stock_picking_backorder_reason: Add some demo data

rousseldenis commented:
@lmignon Could you update your review ?

lmignon reviewed (approved):
Thank you @rousseldenis LGTM (Core review + functional tests)

lmignon reviewed (changes requested):
@rousseldenis The licence must be LGPL since this addon will be mixed with helpdesk.

jbaudoux commented:
@rousseldenis @lmignon I don't like the module name and how the module solves the initial requirement.

Basically, we want to configure on the supplier if he supports backorder in his system or not, and this in order to not keep a reception backorder open in odoo for nothing as you will never receive the missing goods.

So mainly we need to configure a backorder strategy on the supplier.
Then we need to be able to process a reception, make intermediate validation while receiving (like when a pallet is received and you want to be able to put it away before continuing receiving, or like when you want to urgently receive a single product from the reception and continue later) and when everything is received, apply the supplier backorder strategy on the remaining backorder.
When you validate the picking, in standard you get a wizard asking if you want a backorder or not. Instead of asking if we want a backorder or not, you could ask if all received goods have been processed and if not then create a backorder (this could even be configured on the reception picking as backorder strategy to always), otherwise apply the supplier backorder strategy. If no supplier strategy is configured on the supplier, then we even don't need to ask anything, we just apply standard.
The reason basically serves this purpose of asking if all goods have been received.

Besides this, in my opinion, you cannot mirror this feature for sales in the same way. So all the code to try to be generic and cover sales looks useless to me.

I would simplify a lot this module to just keep the basic need regarding supplier and rename the module to reflect that it basically adds a supplier backorder strategy.

rousseldenis commented:
> Besides this, in my opinion, you cannot mirror this feature for sales in the same way. So all the code to try to be generic and cover sales looks useless to me.

We have flows where customer has particular requirement on this, so we need to keep that.

> I would simplify a lot this module to just keep the basic need regarding supplier and rename the module to reflect that it basically adds a supplier backorder strategy.

From my preceding remark, I agree partly on this as this could be renamed 'backorder_partner_stragey' or 'backorder_third_party_strategy' to reflect a module centered on the 'partner' strategy, letting the 'reason' as accessory and not the main functionnality.

rousseldenis committed:
[FIX] stock_backorder_reason: Pass default_pick_ids in context for cancel

Due to https://github.com/odoo/odoo/commit/74a7309ebbaf6d3bc0c3784b7109a95612f67b9f#diff-c225137b15a0c221c12556a79d41a7cb32c52810558fe8206e2007e05118105eR74
pick_ids is used in context for cancel flow.

rousseldenis committed:
[FIX] stock_picking_backorder_reason: Don't show wizard in every picking

rousseldenis committed:
[FIX] stock_picking_backorder_reason: Add view for partners

rousseldenis committed:
[IMP] stock_picking_backorder_reason: Add first french translation

rousseldenis committed:
[IMP] stock_picking_backorder_reason: Allow to cancel backorder transparently

As user does not want to have necessarily a reason wizard to cancel the backorder,
add a parameter that allows to bypass that wizard.

rousseldenis committed:
[FIX] stock_picking_backorder_reason: Use the transparent cancel option as a separate flow

The reason option should not be mandatory to let the transparent cancel flow will occur.

So, let the options for sale and purchase be enabled without the reason one.

rousseldenis committed:
[IMP] stock_picking_backorder_reason: Use Odoo stock method to transparent cancel backorders

Pass 'picking_ids_not_to_backorder' context key when doing picking validation
in order to do transparent backorder cancelation.

rousseldenis commented:
> I would simplify a lot this module to just keep the basic need regarding supplier and rename the module to reflect that it basically adds a supplier backorder strategy.

We have the case for customers too.
